### PR TITLE
[DEVHAS-228] Override local Dockerfile URIs in Component devfile when Dockerfile is external

### PR DESCRIPTION
**Pull request by johnmcollier**
### What does this PR do?:
Updates the Component controller so that when a DockerfileURL is passed in, if the Component devfile contains local Dockerfile URIs, those URIs will be updated to use the fullly qualified DockerfileURL.

Additionally, the CDQ controller has now been updated to no longer return a DockerfileURL if a Dockerfile is found locally within the detected Component, as the Build process will take care of resolving the Dockerfile.

### Which issue(s)/story(ies) does this PR fixes:
https://issues.redhat.com/browse/DEVHAS-228

### PR acceptance criteria:

- [x] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

File: pkg/devfile/devfile.go [hunk 1]
```diff
@@ -250,3 +253,30 @@ func DownloadDevfileAndDockerfile(url string) ([]byte, string, []byte) {
...
+	devfileComponents, err := devfile.GetComponents(common.DevfileOptions{ComponentOptions: common.ComponentOptions{
+		ComponentType: v1alpha2.ImageComponentType,
+	}})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, comp := range devfileComponents {
+		if comp.Image != nil && comp.Image.Dockerfile != nil {
+			// If the DockerfileURI is local (i.e. not a fully qualified URL) overwrite it to be the given Dockerfile URL, otherwise ignore it
```
Comment (maysunfaisal) [inline pkg/devfile/devfile.go:268]: Why not? If Component Source has DockerfileURL, we probably want to overwrite it always?

File: controllers/component_controller.go [hunk 2]
```diff
@@ -308,6 +308,7 @@ func (r *ComponentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 				return ctrl.Result{}, err
 			}
 
+			// If the DockerfileURL is set in the Git source - update any local references to the Dockerfile in the devfile to use the fully qualified URL
```
Comment (maysunfaisal) [inline controllers/component_controller.go:311]: probably want to place this somewhere close to the actual line

File: pkg/devfile/devfile.go (hunk 1 above)
Comment (johnmcollier) [reply pkg/devfile/devfile.go:268]: My thinking was that we wouldn't want to overwrite it, but there be cases where users pass in DockerfileURL to overwrite a fully qualified URL in the devfile that isn't resolvable or maybe private. I'll update it

Comment (johnmcollier) [reply pkg/devfile/devfile.go:268]: Updated.

File: controllers/component_controller.go (hunk 2 above)
Comment (johnmcollier) [reply controllers/component_controller.go:311]: Updated

**Comment by johnmcollier** [comment]
@maysunfaisal Review comments addressed

**Review by maysunfaisal — APPROVED** [review]

**Pull request merged**